[BEAM-4196][SQL] Support complex types in DDL

akedin opened the pull request:
Support complex types in DDL

Supported syntax:

```
CREATE TABLE tableName (
  f_array2 ARRAY<INTEGER>,
  f_array3 ARRAY<ARRAY<INTEGER>>,
  f_map MAP<INTEGER, MAP<VARCHAR, VARCHAR>>,
  f_row ROW( f_int1 INTEGER, f_str1 VARCHAR),
  f_row2 ROW< f_int1 INTEGER, f_str1 VARCHAR>
)
```


------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `./gradlew build` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

akedin commented:
R: @apilloud @kennknowles @XuMingmin @xumingming

apilloud commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlDdlNodes.java:
```diff
@@ -28,14 +27,6 @@
 public class SqlDdlNodes {
   private SqlDdlNodes() {}
 
-  /** Creates a CREATE TABLE. */
```
This annoys me. I don't like this extra indirection in calcite but I also don't like deviating from calcite unnecessaraly. Not much you can do about that.

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
thread leak test fails: https://github.com/apache/beam/pull/5300/   BEAM-4088

kennknowles commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlCliTest.java:
```diff
@@ -41,13 +50,178 @@ public void testExecute_createTextTable() throws Exception {
+        + "name varchar COMMENT 'name', \n"
+        + "age int COMMENT 'age', \n"
+        + "tags MAP<VARCHAR, VARCHAR>, \n"
+        + "nestedMap MAP<INTEGER, MAP<VARCHAR, INTEGER>> \n"
```
<!--new_thread; commit:acf50611c0f2b329b63e3a21ccc9e54db51e5d97; resolved:0-->
Can you also test:

 - maps inside arrays
 - arrays inside maps
 - rows inside arrays
 - rows inside maps
 - maps inside rows
 - arrays inside rows
 - nested nested rows

... etc ...

At some point we might want to use QuickCheck/SmallCheck; that is one of the best ways to test valid syntax.

akedin replied:
Will do

kennknowles commented on sdks/java/extensions/sql/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -153,4 +190,115 @@ SqlDrop SqlDropTable(Span s, boolean replace) :
+        fieldType = SimpleType()
+    )
+    [
+        collectionTypeName = CollectionTypeName()
```
<!--new_thread; commit:acf50611c0f2b329b63e3a21ccc9e54db51e5d97; resolved:0-->
This is to support `INT ARRAY` in that syntax for it? TBH I don't think we need to start with that, even though it is "standard" all the related dialects don't use that syntax.

And I didn't realize JavaCC generated LL(k) parsers. That's unfortunate. So I can see why the grammar has to be factored like this. But what this doesn't support is (INT ARRAY) ARRAY or the no-parens version INT ARRAY ARRAY.

I suppose it is fine to leave as-is, but probably want to file something about the limitation, or tell me that I am wrong. And lots of tests. And also feel free to remove if you don't want the complication.

akedin replied:
Yes this clause is for the postfix array support. I left it mostly because it was already kinda there, just limited to `MULTISETS`, not arrays. We don't support either at the moment, and I would rather remove it for now for consistency

kennknowles commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/parser/BeamSqlParserTest.java:
```diff
@@ -63,7 +64,7 @@ public void testParseCreateTable_withoutType() throws Exception {
         "create table person (\n"
             + "id int COMMENT 'id', \n"
-            + "name varchar(31) COMMENT 'name') \n"
+            + "name varchar COMMENT 'name') \n"
```
<!--new_thread; commit:4a38a26e2bfa0631ef5426455a57802eb81e5e93; resolved:0-->
OK, I commented on the CLI test, but this is the better place for the parser tests.

akedin replied:
Adding QuickCheck tests

kennknowles commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/parser/SqlCreateTable.java:
```diff
@@ -67,23 +72,24 @@
   public List<SqlNode> getOperandList() {
-    return ImmutableNullableList.of(name, columnList, type, comment, location, tblProperties);
+    throw new UnsupportedOperationException(
+        "Getting operands CREATE TABLE is unsupported at the moment");
```
<!--new_thread; commit:4a38a26e2bfa0631ef5426455a57802eb81e5e93; resolved:1-->
Just curious - where is it used? You can't really do much except copy the node if you don't know what kind of node it is, so just asking for my own education how Calcite (or BeamSQL) uses this.

akedin replied:
It's mostly used by implementations of the `SqlCall` to access their parameters. `SqlCreate` itself doesn't really need it but we might (or might not) need something like this for JDBC integration, depending on how `CREATE TABLE` parsing will be implemented there

akedin commented:
Updated:
* removed postfix array support;
* added (`BeamDdlNestedTypesTest`)[https://github.com/apache/beam/pull/5276/files#diff-794c03679a02b7349f5070def998ce7a] which tests things with QuickCheck. Examples:

```
create table tablename ( fieldName MAP<DOUBLE, ARRAY<ARRAY<ARRAY<MAP<BOOLEAN, MAP<DECIMAL, TINYINT>>>>>> ) TYPE 'text' LOCATION '/home/admin/person'

create table tablename ( fieldName MAP<BOOLEAN, MAP<FLOAT, ARRAY<BOOLEAN>>> ) TYPE 'text' LOCATION '/home/admin/person'

create table tablename ( fieldName ARRAY<ARRAY<FLOAT>> ) TYPE 'text' LOCATION '/home/admin/person'


create table tablename ( fieldName MAP<INTEGER, MAP<INTEGER, MAP<INTEGER, MAP<SMALLINT, ROW<field_0 INTEGER,field_1 TIMESTAMP,field_2 TIMESTAMP,field_3 TIMESTAMP>>>>> ) TYPE 'text' LOCATION '/home/admin/person'


create table tablename ( fieldName MAP<BOOLEAN, ROW<field_0 INTEGER,field_1 DECIMAL,field_2 VARCHAR>> ) TYPE 'text' LOCATION '/home/admin/person'



```

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
run java precommit

akedin commented:
whoa it's green

kennknowles commented:
Awesome. That's next-level test assurance.
